From 6ea64c4b892e68651aba10fb6cff0b2d00eb9302 Mon Sep 17 00:00:00 2001
From: Suren Baghdasaryan <surenb@google.com>
Date: Wed, 11 Sep 2024 10:03:34 -0700
Subject: [PATCH] lmkd: fix handling of EPOLLHUP for pidfd

Recent kernel change [1] cause pidfd_wait() to receive EPOLLHUP when the
task exits. Current LMKD implementation expects to receive EPOLLHUP only
when socket connection gets dropped, therefore it gets confused by this
new kernel behavior. Adjust LMKD handling of EPOLLHUP to detect the case
when this event is generated by pidfd.

[1] https://lore.kernel.org/all/20240202131226.GA26018@redhat.com/

Bug: 352286227
Change-Id: Ibcf349ee3cc73551541d64975f0292d53c41d5c2
Signed-off-by: Suren Baghdasaryan <surenb@google.com>
---
 lmkd.cpp | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/lmkd.cpp b/lmkd.cpp
index 79adc22..b19aed6 100644
--- a/lmkd.cpp
+++ b/lmkd.cpp
@@ -3643,11 +3643,15 @@ static void mainloop(void) {
          */
         for (i = 0, evt = &events[0]; i < nevents; ++i, evt++) {
             if ((evt->events & EPOLLHUP) && evt->data.ptr) {
-                ALOGI("lmkd data connection dropped");
                 handler_info = (struct event_handler_info*)evt->data.ptr;
-                watchdog.start();
-                ctrl_data_close(handler_info->data);
-                watchdog.stop();
+                if (handler_info->handler == kill_done_handler) {
+                    call_handler(handler_info, &poll_params, evt->events);
+                } else {
+                    ALOGI("lmkd data connection dropped");
+                    watchdog.start();
+                    ctrl_data_close(handler_info->data);
+                    watchdog.stop();
+                }
             }
         }
 
-- 
2.49.0

